Incompatible XSMP cmake option

Hello !

I am using xsmp-sdk (main branch, commit da488bcee785c9176b5ddf433df819bda9eb08c9) and the xsmp-modeler VSCode extension.

In my current environment, I need to enable the install rule (with `-DXSMP_ENABLE_INSTALL=ON` so that, targets can be exported and used).
During my project development, I usually try to install into specific directory for packaging using the `--prefix` like:
`cmake --install build --prefix /tmp/test`

However, this fails as it tries to install `ecss_smp.cpython` into `/usr/lib/python3.9` (which requires admin permissions and ignores the `--prefix` argument).
I tried disabling the python bindings with option (`-DXSMP_BUILD_PYTHON_BINDINGS=OFF`) but this failed too as the variable `Python_SITELIB` is not defined anymore. This gives the following error:
```
CMake Error at build/Debug/models/xsmp-sdk/cmake_install.cmake:62 (file):
  file INSTALL given no DESTINATION
Call Stack (most recent call first):
  build/Debug/models/cmake_install.cmake:77 (include)
  build/Debug/cmake_install.cmake:47 (include)
```

It seems to me that this combinaison of option is invalid `-DXSMP_ENABLE_INSTALL=ON` and `-DXSMP_BUILD_PYTHON_BINDINGS=OFF`, is this expected?

Thank you for your time,
Umut

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -289,6 +289,46 @@ target_include_directories(
 # --------------------------------------------------------------------
 if(XSMP_BUILD_PYTHON_BINDINGS)
     find_package(Python COMPONENTS Interpreter Development)
+
+    if(SKBUILD STREQUAL "2")
+        # the python modules are at the root of the wheel, the libraries in <root>/${CMAKE_INSTALL_LIBDIR}
+        set(_xsmp_pythondir ".")
+    else()
+        # The default install directory of the python modules is computed as a path
+        # relative to the install prefix, so that
+        # `cmake --install <dir> --prefix <prefix>` is honored.
+        # It can be overridden with an absolute path (e.g. ${Python_SITELIB}) to
+        # install the modules outside of the install prefix.
+        if(NOT DEFINED CACHE{XSMP_INSTALL_PYTHONDIR})
+            execute_process(
+                COMMAND "${Python_EXECUTABLE}" -c
+                        "import os, sysconfig; print(os.path.relpath(sysconfig.get_path('purelib'), sysconfig.get_path('data')).replace(os.sep, '/'))"
+                OUTPUT_VARIABLE _xsmp_default_pythondir
+                OUTPUT_STRIP_TRAILING_WHITESPACE
+                ERROR_QUIET
+            )
+            if(NOT _xsmp_default_pythondir AND Python_VERSION_MAJOR)
+                set(_xsmp_default_pythondir "${CMAKE_INSTALL_LIBDIR}/python${Python_VERSION_MAJOR}.${Python_VERSION_MINOR}/site-packages")
+            endif()
+            set(XSMP_INSTALL_PYTHONDIR "${_xsmp_default_pythondir}" CACHE STRING
+                "Directory where the python modules are installed (relative to CMAKE_INSTALL_PREFIX, or absolute)")
+        endif()
+        set(_xsmp_pythondir "${XSMP_INSTALL_PYTHONDIR}")
+    endif()
+
+    # RPATH of the ecss_smp module: the path of the libraries it links, relative
+    # to its own install directory
+    if(IS_ABSOLUTE "${_xsmp_pythondir}" OR IS_ABSOLUTE "${CMAKE_INSTALL_LIBDIR}")
+        set(_xsmp_python_rpath "${CMAKE_INSTALL_FULL_LIBDIR}")
+    else()
+        file(RELATIVE_PATH _xsmp_python_rpath "/${_xsmp_pythondir}" "/${CMAKE_INSTALL_LIBDIR}")
+        if(APPLE)
+            set(_xsmp_python_rpath "@loader_path/${_xsmp_python_rpath}")
+        else()
+            set(_xsmp_python_rpath "$ORIGIN/${_xsmp_python_rpath}")
+        endif()
+    endif()
+
     FetchContent_Declare(
       pybind11
       GIT_REPOSITORY https://github.com/pybind/pybind11.git
@@ -300,10 +340,8 @@ if(XSMP_BUILD_PYTHON_BINDINGS)
     pybind11_add_module(Python src/python/ecss_smp.cpp)
     set_target_properties(Python PROPERTIES OUTPUT_NAME "ecss_smp")
 
-    if(APPLE)
-        set_target_properties(Python PROPERTIES INSTALL_RPATH "@loader_path/${CMAKE_INSTALL_LIBDIR}")
-    elseif(UNIX)
-        set_target_properties(Python PROPERTIES INSTALL_RPATH "$ORIGIN/${CMAKE_INSTALL_LIBDIR}")
+    if(UNIX)
+        set_target_properties(Python PROPERTIES INSTALL_RPATH "${_xsmp_python_rpath}")
     endif()
 
     add_library(Xsmp::Python ALIAS Python)
@@ -316,8 +354,8 @@ if(XSMP_BUILD_PYTHON_BINDINGS)
     target_compile_options(Python PRIVATE ${XSMP_COMPILE_OPTIONS})
     if(SKBUILD STREQUAL "2")
         install(TARGETS Python DESTINATION .)
-    else()
-        install(TARGETS Python DESTINATION ${Python_SITELIB}/)
+    elseif(XSMP_ENABLE_INSTALL)
+        install(TARGETS Python DESTINATION "${XSMP_INSTALL_PYTHONDIR}")
     endif()
 endif(XSMP_BUILD_PYTHON_BINDINGS)
 
@@ -630,7 +668,7 @@ if(XSMP_ENABLE_INSTALL)
             FILES_MATCHING
             PATTERN "*.h"
     )
-    if(NOT SKBUILD STREQUAL "2")
+    if(XSMP_BUILD_PYTHON_BINDINGS AND NOT SKBUILD STREQUAL "2")
         install(DIRECTORY "python/xsmp"
                           "python/xsmp_event_manager"
                           "python/xsmp_link_registry"
@@ -638,7 +676,7 @@ if(XSMP_ENABLE_INSTALL)
                           "python/xsmp_resolver"
                           "python/xsmp_scheduler"
                           "python/xsmp_time_keeper"
-                DESTINATION "${Python_SITELIB}"
+                DESTINATION "${XSMP_INSTALL_PYTHONDIR}"
                 FILES_MATCHING
                 PATTERN "*.py"
         )

--- a/README.md
+++ b/README.md
@@ -29,6 +29,46 @@
 - **Enhanced Productivity**: Using these two projects together accelerates development and reduces errors.
 - **Consistency and Standardization**: Ensures adherence to established standards and practices.
 
+## Building and installing
+
+```bash
+cmake -S . -B build -DCMAKE_BUILD_TYPE=Release
+cmake --build build -j
+# the install rule is enabled by default when xsmp-sdk is the top level project
+cmake --install build --prefix /path/to/install
+```
+
+### CMake options
+
+| Option | Default | Description |
+| --- | --- | --- |
+| `XSMP_BUILD_TESTS` | `ON` for the top level project | Build the unit tests. |
+| `XSMP_BUILD_PYTHON_BINDINGS` | `ON` | Build the `ecss_smp` Python module and install the Python packages. |
+| `XSMP_BUILD_PACKAGE` | `ON` for the top level project | Build the CPack package. |
+| `XSMP_BUILD_EXAMPLES` | `OFF` | Build the examples. |
+| `XSMP_ENABLE_INSTALL` | `ON` for the top level project | Enable the install rule, required to export the targets and use `find_package(xsmp-sdk)`. |
+| `XSMP_ENABLE_CODECOVERAGE` | `OFF` | Enable the code coverage support. |
+| `XSMP_BUILD_WITH_WARNINGS` | `OFF` | Enable all compiler warnings. |
+| `XSMP_INSTALL_PYTHONDIR` | `<libdir>/python<X.Y>/site-packages` | Directory where the Python modules are installed. |
+
+### Installing the Python modules
+
+The `ecss_smp` module and the Python packages (`xsmp`, `xsmp_logger`, ...) are installed in
+`XSMP_INSTALL_PYTHONDIR`, whose default value is deduced from the `site-packages` layout of the
+Python interpreter found at configuration time (e.g. `lib/python3.12/site-packages`).
+
+- A **relative** path (the default) is resolved against the install prefix, so
+  `cmake --install build --prefix /tmp/test` installs everything under `/tmp/test`. Add
+  `/tmp/test/lib/python<X.Y>/site-packages` to your `PYTHONPATH` to use them.
+- An **absolute** path installs the modules outside of the install prefix, e.g. to install them
+  directly in the `site-packages` directory of the current interpreter:
+  ```bash
+  cmake -S . -B build -DXSMP_INSTALL_PYTHONDIR="$(python3 -c 'import sysconfig; print(sysconfig.get_path("purelib"))')"
+  ```
+
+Building with `-DXSMP_BUILD_PYTHON_BINDINGS=OFF` installs no Python module at all; only the C++
+libraries, the headers and the CMake configuration files are installed.
+
 ## Documentation
 
 Comprehensive documentation is available [here](https://ThalesGroup.github.io/xsmp-sdk/). We encourage you to read it to get the most out of the XSMP SDK.

